File: code/graphs/dijkstra.cpp
// Lazy Dijkstra: a min-heap of (dist, vertex) pairs. Settle a vertex the moment
// it POPS as the minimum; skip stale pops; relax its out-edges and push any
// improved neighbor. Needs all weights >= 0.
vector<long> dijkstra(WGraph& g, int s) {
    vector<long> dist(g.V, LONG_MAX);
    vector<bool> settled(g.V, false);
    priority_queue<pair<long,int>, vector<pair<long,int>>,
                   greater<>> pq;              // min-heap keyed by distance
    dist[s] = 0;
    pq.push({0, s});
    while (!pq.empty()) {
        int u = pq.top().second; pq.pop();     // nearest unsettled
        if (settled[u]) continue;              // stale entry — skip
        settled[u] = true;                     // dist[u] is now FINAL
        for (Edge e : g.adj[u])
            if (!settled[e.to] && dist[u] + e.w < dist[e.to]) {
                dist[e.to] = dist[u] + e.w;     // relax: a cheaper route to e.to
                pq.push({dist[e.to], e.to});    // push, don't decrease-key
            }
    }
    return dist;
}

File: code/graphs/dijkstra.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

// Lazy Dijkstra. BinaryHeap is a MAX-heap, so wrap each (dist, vertex) in
// Reverse to pop the SMALLEST distance. Settle on pop; skip stale pops.
fn dijkstra(g: &WGraph, s: usize) -> Vec<i64> {
    let n = g.adj.len();
    let mut dist = vec![i64::MAX; n];    // dist[v] = best-known cost s->v
    let mut settled = vec![false; n];    // settled[v] = dist[v] is final
    let mut pq = BinaryHeap::new();
    dist[s] = 0;
    pq.push(Reverse((0i64, s)));
    while let Some(Reverse((_d, u))) = pq.pop() {   // nearest unsettled
        if settled[u] { continue; }                 // stale entry — skip
        settled[u] = true;                          // dist[u] is now FINAL
        for &(v, w) in &g.adj[u] {
            if !settled[v] && dist[u] + w < dist[v] {
                dist[v] = dist[u] + w;              // relax: cheaper route to v
                pq.push(Reverse((dist[v], v)));
            }
        }
    }
    dist
}
